tests/realtikvtest/pessimistictest/pessimistic_test.go

```diff
@@ -1332,6 +1332,82 @@ func TestPessimisticLockNonExistsKey(t *testing.T) {
 	tk1.MustExec("rollback")
 }
 
+func TestUnionAllPointGetLockBlockInsert(t *testing.T) {
+	store, clean := realtikvtest.CreateMockStoreAndSetup(t)
+	defer clean()
+
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("create table t (id int primary key, v int)")
+
+	type test struct {
+		sql         string
+		shouldBlock bool
+
+		concurrentSQL string
+	}
+
+	cases := []test{
+		{
+			"select * from t where id=1 for update union all select * from t where id = 2 for update",
+			true,
+			"insert into t values (1, 1), (2, 2)",
+		},
+		{
+			"select * from t where id=1 for update union all select * from t where id = 2",
+			false,
+			"insert into t values (2, 2)",
+		},
+		{
+			"select * from t where id=1 union all select * from t where id = 2 for update",
+			false,
+			"insert into t values (1, 1)",
+		},
+	}
+
+	for _, c := range cases {
+		ch := make(chan string, 1)
+
+		tk.MustExec("truncate t")
+		tk.MustExec("begin pessimistic")
+		tk.MustExec(c.sql)
+
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+
+			tk2 := testkit.NewTestKit(t, store)
+			tk2.MustExec("use test")
+			tk2.MustExec(c.concurrentSQL)
+
+			ch <- "done"
+			wg.Done()
+		}()
+
+		if c.shouldBlock {
+			select {
+			case <-ch:
+				panic("Should receive nothing")
+			case <-time.After(time.Second * 1):
+				break
+			}
+		} else {
+			select {
+			case <-ch:
+				break
+			case <-time.After(time.Second * 1):
```

Comment:
This line may lead to fail when the environment executing test cases runs slowly, I think this is one of the reasons that test cases run unstably on CI.
But I have no good idea.